File: Data Structures & Algorithms/valid-tree/submission-5.js
class Solution {
    /**
     * @param {number} n
     * @param {number[][]} edges
     * @returns {boolean}
     */
    validTree(n, edges) {
        //hint, what are edges? and, we cant go back where we came from

        let seen = new Set(), map = new Map();

        for(let i = 0; i < n; i++){
            map.set(i, []);
        }

        for(const [l, r] of edges){
            map.get(l).push(r);
            map.get(r).push(l);
        }

        function dfs(src, pr){
            if(seen.has(src)) return false;
            seen.add(src);

            for(const l of map.get(src)){
                if(pr === l) continue;
                if(!dfs(l, src)) return false;
            }
            //seen.delete(src);
            return true;
        }

        dfs(0, -1); //-1 for no parent

        for(let i = 0; i < n; i++){
            if(!seen.has(i)) return false;
        }

        return edges.length === n - 1;
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-5.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = {}
        edgess = 0

        for i in range(n):
            adj[i] = []

        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)
        
        seen = set()

        def dfs(node, parent):

            if node in seen:
                return
            
            seen.add(node)
            
            for edge in adj[node]:
                if not edge == parent and not edge in seen:
                    dfs(edge, node)
            
        
        for i in range(n):
            if not i in seen:
                dfs(i, -1)
                edgess += 1
        
        return edgess
            

            